marketplace/deployer_util/validate.py

```diff
@@ -0,0 +1,185 @@
+#!/usr/bin/env python2
+#
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import json
+import yaml
+
+from argparse import ArgumentParser
+from bash_util import Command
+from dict_util import deep_get
+from yaml_util import load_resources_yaml
+from yaml_util import load_yaml
+
+DEPLOY_INFO_ANNOTATION = "marketplace.cloud.google.com/deploy-info"
+GOOGLE_MARKETPLACE = "x-google-marketplace"
+
+_PROG_HELP="""
+Validates and test the deployer image.
+"""
+
+class ValidationException(Exception):
+  pass
+
+def main():
+  parser = ArgumentParser(description=_PROG_HELP)
+  parser.add_argument('--metadata', help='Path to metadata file for validation. If it is not present, the validation on the metadata file will be skipped.',)
+  args = parser.parse_args()
+
+  all_resources, application = expand_and_load_resources()
+
+  metadata = None
+  if args.metadata:
+    with open(args.metadata, "r") as metadatafile:
+      metadata = json.loads(metadatafile.read())
+  
+  if metadata:  
+    validate_metadata(metadata, application)
+
+  schema = load_yaml("/data/schema.yaml")
+  validate_images(schema, all_resources)
+
+
+def validate_images(schema, resources):
+  declared_images = set([deep_get(param, "default") for key, param in schema["properties"].iteritems() 
+    if deep_get(param, GOOGLE_MARKETPLACE, "type") == "IMAGE"])
+
+  used_images = []
+  for resource in resources:
+    used_images.extend(deep_find(resource, "image"))
```

Comment:
Not entirely sure about blanketing all attributes with `image` names. Maybe we should handle known workload resource types only.
- Pod
- Deployment
- StatefulSet
- DaemonSet
- Job
- CronJob
- ReplicaSet
- ReplicationController

See [k8s doc](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/). There isn't a landing page that lists all the workloads! Please look at the content tree on the LHS.

Reply:
This is tricky. I would opt for false fails rather than false successes.

a) If we are too specific, we run the risk of skipping images.
b) If we do a catch all, we run the risk of doing extra validation.

I prefer fixing a bug in (b) rather than (a). If (b) causes trouble to us, that will be a partner complaint. If (a) causes trouble, that will probably be complains from multiple customers.